Support an already opened file as argument to WahWah::open

obskyr opened the pull request:
I wish to be able to use WahWah in conjunction with [Down](https://github.com/janko/down) in order to read the tags from audio files hosted on the internet without downloading the entire files, as following:

```ruby
require "wahwah"
require "down"

stream = Down.open "https://github.com/obskyr/obscuriosities/raw/70a37bc126b8fe160ec2d137bec97bf69b2461dc/episodes/Episode%202%20-%20Mon%20Dieu,%20It's%20Jumping%20Flash%203-finished.ogg"
tags = WahWah.open stream
```

Given that the file in question isn't an ID3v1 file (which keeps the metadata at the *end* of the file; mamma mia), this would only require downloading the first kilobyte or so of the file! Handy, right?

However, as it stands, WahWah only supports supplying a path to a local file to `open`, since it uses the filename's extension to determine the filetype. The patch in this pull request migrates filetype detection to using the file signature instead (e.g. `ID3` for ID3v2; `fLaC` for FLAC…), and allows you to use already opened file objects (and file-like objects!) with WahWah. As an added bonus, it also enables WahWah to work with media files with unconventional extensions!

Since the rest of the code seemed to be more tightly coupled with filetype than with tag formats (e.g. `Mp3Tag` instead of, say, `Id3v2Tag`), I left that be, and extensions are still used internally at one point – but that's required on some level to keep `WahWah::support_formats` and thus preserve API compatibility, either way.

I've ensured that all the tests pass, as well as added a new test for using an already-opened file. Let me know if you'd like any changes, and I'd be happy to make them!

obskyr commented:
Should this pull request be merged, I have another one in the pipes that would allow getting Ogg metadata while lazy-loading the duration (since one needs to read to the end for that – what a bizarre downside of the format) – but it only really makes sense with this one as a base, so I will wait for this one to be processed first.

I've written documentation (that is to say, a short section in the readme) for the new features as well – that, too, would be part of the second PR.

aidewoode commented:
Thanks for your PR, it's so great that wahwah could accept IO like object as argument. It's going to be very useful. I will review your code later because I don't have much time right now. But I promise it won't take too long.

obskyr commented:
Looks like I updated the formatting of `helper_test.rb` after testing, and the linter didn't like that – apparently, it doesn't want you to align values in a hash. I've updated the PR – all the tests should pass now!

aidewoode commented on lib/wahwah/mp3_tag.rb:
```diff
@@ -50,11 +50,21 @@ def parse
-
-      return id3_v2_tag if id3_v2_tag.valid?
-      id3_v1_tag if id3_v1_tag.valid?
+      @file_io.seek(0)
```
Can use rewind instead of seek(0)

aidewoode commented on lib/wahwah/mp3_tag.rb:
```diff
@@ -50,11 +50,21 @@ def parse
-      id3_v1_tag if id3_v1_tag.valid?
+      @file_io.seek(0)
+      signature = @file_io.read(6)
+      @file_io.seek(0)
```
No need to rewind twice, because ID3::V2 and ID3:V1 will seek to specific position when parse.

aidewoode commented on lib/wahwah/tag.rb:
```diff
@@ -27,25 +27,30 @@ class Tag
     )
 
     def initialize(file)
```
Since we already have handle params in WhaWah.open, this initialize method should only accept IO as params.

obskyr replied:
I allowed it to accept both because I didn't want to encroach on too much code, and the tests often use, for example, `Mp3Tag.new "example.mp3"`. Would you like me to implement this and rewrite the tests to use `WahWah::open` (or `XxxTag::new` with an IO, when testing specific `Tag` subclasses)?

aidewoode replied:
I used to allow `Tag` class to accept IO like params mainly because in some tests I need to pass IO object to make test easy. So now we don't need to handle file path in `Tag` initialize method anymore, I think just accepting the IO is better. And for testing specific tag subclasses, you can `File.open` file path can pass IO object to the class. I know there are a lot of places in the test need to change. It's very tedious. I hope you don't mind.

obskyr commented on lib/wahwah/helper.rb:
```diff
@@ -25,8 +25,31 @@ def self.split_with_terminator(string, terminator_size)
 
-    def self.file_format(file_path)
-      File.extname(file_path).downcase.delete(".")
+    def self.file_format(file)
```
Do we *want* to determine the format based on the extension? It's generally a less correct way to determine a filetype. Is it for backward compatibility…?

aidewoode replied:
Yes, I know the file extension may be wrong, but this is the user's mistake. I think it is the user's responsibility to make sure the file extension is correct. If user provide the extension name, the program should treat the file as the format, if can not parse the file, then the user should know there are some problems in the file. So I think it's reasonable to first get format from file extension if we could. Also it's faster to get format from extension name.

obskyr commented on lib/wahwah/helper.rb:
```diff
@@ -25,8 +25,31 @@ def self.split_with_terminator(string, terminator_size)
+      # M4A is checked for first, since MP4 files start with a chunk size -
+      # and that chunk size may incidentally match another signature.
+      # No other formats would reasonably have "ftyp" as the next for bytes.
+      return "m4a" if signature[4...8] == "ftyp".b && signature[8...12] == "M4A ".b
```
I figured this was more readable, but I can change it! (Though it should be `signature[4...12] == "ftypM4A "`, with the space.)

obskyr commented:
@aidewoode I've addressed all the review items! This should be ready to merge now – take a look at the changes and see, I suppose.

After that I have another pull request coming up that suggests a slight API change (making tags that are liable to appear toward the end of the file lazy-loaded so that streamed files don't automatically and eagerly download the whole file), so I'd recommend not releasing this particular commit as a numbered version, as that'd constitute a breaking change.

aidewoode commented:
Thanks for your great contribution and hard work.  Although the code still has a few issues, because I don't want to interrupt you too much for the PR, so I merged the code, and for those little issues I will push a commit to fix and you don't need to worry. I do want to mention some issues in the code, though, because I think those will make you more familiar with Ruby.

- `private_class_method` does not make the class method private unless you pass the method name. see: https://stackoverflow.com/questions/4952980/how-to-create-a-private-class-method
- The default mode of `File.open` is `r`, the read-only mode. And `IO#read` will read bytes without any conversion (binary mode). So I think you don't need to add `rb` to all tests.
